R/shiny_prerendered.R

```diff
@@ -0,0 +1,386 @@
+  pattern <- paste0('<script type="application/shiny-prerendered" data-context="', context, '">')
+  matches <- regmatches(html_lines, regexec(pattern, html_lines))
+
+  # extract the code within the contexts
+  in_context <- FALSE
+  context_lines <- character()
+  for (i in 1:length(matches)) {
+    if (length(matches[[i]]) > 0) {
+      in_context <- TRUE
+      next
+    }
+    else if (in_context && identical(html_lines[[i]], "</script>")) {
+      in_context <- FALSE
+    }
+    if (in_context)
+      context_lines <- c(context_lines, html_lines[[i]])
+  }
+  context_lines
+}
+
+
+# Gather shiny_prerendred contexts and append them as script tags to
+# the passed file
+shiny_prerendered_append_contexts <- function(runtime, file, encoding) {
+
+  # collect contexts
+  shiny_prerendered_contexts <- knit_meta_reset(class = "shiny_prerendered")
+  if (length(shiny_prerendered_contexts) > 0) {
+
+    # validate we are in runtime: shiny_prerendered
+    if (!is_shiny_prerendered(runtime)) {
+      stop("The code within this document requires runtime: shiny_prerendered",
+           call. = FALSE)
+    }
+
+    # open the file
+    con <- file(file, open="at", encoding = encoding)
+    on.exit(close(con), add = TRUE)
+
+    # append the contexts as script tags
+    for (context in shiny_prerendered_contexts) {
+      lines <- c(paste0('<script type="application/shiny-prerendered" ',
+                        'data-context="', context$name ,'">'),
+                 context$code,
+                 '</script>')
+      writeLines(lines, con = con)
+    }
+  }
+}
+
+
+# Prerendred data_dir for a given Rmd input file
+shiny_prerendered_data_dir <- function(input, create = FALSE) {
+  data_dir <- paste0(tools::file_path_sans_ext(input), "_data")
+  if (create && !dir_exists(data_dir))
+    dir.create(data_dir)
+  data_dir
+}
+
+
+# Load prerendred data into the server environment. There are two
+# reasons we load data based on a previously generated index file:
+#
+# (1) We don't remove all of the previously generated .RData files
+#     at the beginning of a render. Specifically, we don't remove
+#     .RData files that were created within a chunk w/ caching
+#     enabled, since that .RData won't necessarily be re-created on
+#     the next render. This means that we could be left with "stale"
+#     .RData files from cached chunks. Reading based on the list
+#     ensures that we only read .RData for chunks that were
+#     included in the last rendered document.
+#
+# (2) We want to load data into the server environemnt in the
+#     exact same chunk order that it appears in the document.
+#
+shiny_prerendered_data_load <- function(input_rmd, server_envir) {
+  # check for data_dir
+  data_dir <- shiny_prerendered_data_dir(input_rmd)
+  if (dir_exists(data_dir)) {
+    # read index of data files
+    index_file <- shiny_prerendred_data_chunks_index(data_dir)
+    if (file.exists(index_file)) {
+      rdata_files <- readLines(index_file)
```

Comment:
I believe so -- but it's not just a matter of matching encoding. The names of the files may contain characters that are not even representable even in the system encoding. For instance on Windows the system encoding is not likely to be UTF-8 and chunk names may include e.g. multi-byte characters that are not representable in the system encoding.